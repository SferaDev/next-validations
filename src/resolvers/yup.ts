import * as Yup from 'yup';

export function yupResolver<T extends Yup.AnyObjectSchema>(schema: T) {
  return {
    validate: (data: unknown) => schema.validateSync(data),
  };
}
