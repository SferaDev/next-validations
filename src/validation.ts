import type { LiteralUnion } from "type-fest";

import { fastestValidatorResolver } from "./resolvers/fastest-validator";
import { joiResolver } from "./resolvers/joi";
import { yupResolver } from "./resolvers/yup";
import { zodResolver } from "./resolvers/zod";
import { valibotResolver } from "./resolvers/valibot";

export type SCHEMA_TYPE = LiteralUnion<
  | "Yup"
  | "FastestValidator"
  | "Joi"
  | "Zod",
  | "Valibot"
  | string
>;

export function createResolver(type: SCHEMA_TYPE, schema: any) {
  switch (type) {
    case "Yup":
      return yupResolver(schema);
    case "FastestValidator":
      return fastestValidatorResolver(schema);
    case "Joi":
      return joiResolver(schema);
    case "Zod":
      return zodResolver(schema);
    case "Valibot":
      return valibotResolver(schema);


    default:
      throw new Error(`Does not support ${type} validation yet!`);
  }
}
