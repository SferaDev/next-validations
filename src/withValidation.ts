import { type NextApiRequest, type NextApiResponse } from "next";

import { createResolver } from "./validation";
import type { SCHEMA_TYPE } from "./validation";

type NextHandler = (err?: Error) => void;

type ValidationHoF = {
  type: SCHEMA_TYPE;
  mode?: "body" | "query" | "headers";
  schema: unknown;
};

export function withValidation({
  type,
  schema,
  mode = "query",
}: ValidationHoF) {
  return withValidations([{ type, schema, mode }]);
}

export function withValidations(validations: ValidationHoF[]) {
  return (
    handler?: (req: NextApiRequest, res: NextApiResponse) => any,
  ) => async (
    req: NextApiRequest,
    res: NextApiResponse,
    next?: NextHandler,
  ) => {
      try {
        validations.forEach((validation) => {
          const resolver = createResolver(validation.type, validation.schema);
          resolver.validate(req[validation.mode ?? "query"]);
        });

        if (next) {
          next(); return;
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        res.status(400).send(error);
      }
    };
}
