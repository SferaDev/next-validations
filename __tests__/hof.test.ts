import * as yup from "yup";
import { withValidation, withValidations } from "../src";

const schema = yup.object().shape({
  name: yup.string().required(),
});

describe("withValidation", () => {
  it("should return function", () => {
    expect(withValidation({
      schema,
      type: "Yup",
      mode: "query",
    })).toMatchSnapshot();
  });
});

describe("withValidations", () => {
  it("should return function", () => {
    expect(withValidations([{
      schema,
      type: "Yup",
      mode: "body",
    }])).toMatchSnapshot();
  });
});
