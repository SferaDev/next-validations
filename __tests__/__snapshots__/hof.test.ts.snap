// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`withValidation > should return function 1`] = `[Function]`;

exports[`withValidations > should return function 1`] = `[Function]`;
